feat(robot-server): handle new offset locations in /labwareOffsets

sfoster1 opened the pull request:
Adds the new format for labware offset locations from #17363 to the HTTP API `/labwareOffsets` endpoint.

The endpoint now accepts and returns labware offset locations as these sequences. The filter API is for now unchanged, and still works properly to filter out or in offsets based on their core features or on their locations.

The types are defined separately from the protocol engine because we'll be shortly adding a concept of a "general offset" to this API, and that is best modeled as another heterogenous union entry, which means the types will start to diverge.

This comes along with a sql schema migration; the offset locations are normalized out into an offset table with a foreign key into the main offset table. All content from the offset table is migrated to the new one in the standard persistence directory migration api.

## Reviews
- does the sql look good?
- do the tests look appropriate? some have expanded
- do the migrations look good?

## Testing
- [x] put this on a machine that has offset data and make sure it migrates and comes up with the data preserved

sfoster1 committed:
here is a bad server integration

sfoster1 committed:
fixups

sfoster1 committed:
lint and persistence test fixes

sfoster1 committed:
fixes and database improvements

sfoster1 committed:
more

sfoster1 committed:
hey it works now

[pounding knife and fork on table] CTE! CTE! CTE!

sfoster1 committed:
now with fewer ctes

sfoster1 committed:
migration tests

sfoster1 committed:
fix models for list of enums

sfoster1 committed:
fix negative filters

sfoster1 committed:
little formats

sfoster1 committed:
refactor for complexity lint

mjhuff commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
```diff
@@ -0,0 +1,134 @@
+        copy_contents(source_dir=source_dir, dest_dir=dest_dir)
+
+        # First we create the new version of our labware offsets table and sequence table
+        with sql_engine_ctx(
```
Would it be safer to wrap all these operations in a single transaction?

mjhuff commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
Just curious, do we create any kind of db backup when performing migrations?

sfoster1 replied:
Yes, or rather, we leave the old one. Before the code in this file runs, we actually make a copy of the entire persistence directory (including the database file) to a new directory, so this operation isn't actually touching the pre-migration database.

mjhuff reviewed (approved):
That's some fun SQL. Thank you for doing this! I did test it on a robot and can confirm it works, see
[migration_test.json](https://github.com/user-attachments/files/18623097/scratch_28.json)

Couple Qs for my own edifcation.

ddcc4 commented on robot-server/robot_server/labware_offsets/_search_query_builder.py:
```diff
@@ -0,0 +1,177 @@
+    def do_active_filter(self, active: bool) -> Self:
+        """Filter to only active=active rows."""
+        self._current_base_filter_statement = self._current_base_filter_statement.where(
+            labware_offset_table.c.active == True  # noqa: E712
```
Drive-by comment: What is the point of the `active` argument then?
And what is the docstring referring to when it says `only active=active rows`?

sfoster1 replied:
in the comment, the first active is the column name and the second active is the argument name. could stand to be a little clearer - it's saying it filters to only rows whose active column matches the parameter.

Uhhh and that looks like a typo that it filters it to `c.active == True` instead of `c.active == active`

sfoster1 committed:
wrong activity

sfoster1 committed:
this is after all the point of transactions

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_10.py:
```diff
@@ -0,0 +1,420 @@
+)
+
+labware_offset_location_sequence_components_table = sqlalchemy.Table(
+    "labware_offset_sequence_components",
```
Nitpick: theoretically, our naming convention for table names is singular (`labware_offset_sequence_component`).

This hasn't been followed well in all of our tables and I don't know if it's a battle worth fighting.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
Instead of creating a new schema 10 and adding a 9->10 migration, I think we should overwrite the existing schema 9. The existing schema 9 was a first pass that's only existed on dev robots for 3 weeks; it doesn't seem worthwhile for us to carry forward indefinitely.

We can change `LATEST_VERSION_DIRECTORY` from `"9"` to `"9.1"` so that our dev robots re-run `v8_to_v9.py`. Then they'll all cleanly migrate to your new schema, albeit with ~3-week-old data.

I guess it depends on whether we can do this before the first-pass schema 9 ends up in an internal release.

sfoster1 replied:
I think probably we cannot, unfortunately.

SyntaxColoring commented on robot-server/robot_server/labware_offsets/store.py:
```diff
@@ -149,36 +128,114 @@ def __init__(self, bad_offset_id: str) -> None:
+                primary_component_value=component.addressableAreaName,
+                component_value_json=component.model_dump_json(),
+            )
+        else:
+            print(f"ISINSTANCE FAILED: {component}")
+            pass
```
`print()` left over from debugging?

Can this be a `typing_extensions.assert_never()` instead of a `pass`?

sfoster1 replied:
yes and yes

SyntaxColoring commented on api/src/opentrons/protocol_engine/types/labware_offset_location.py:
```diff
@@ -48,12 +48,16 @@ class OnAddressableAreaOffsetLocationSequenceComponent(BaseModel):
 
 
-LabwareOffsetLocationSequenceComponents = (
+LabwareOffsetLocationSequenceComponentsUnion = (
     OnLabwareOffsetLocationSequenceComponent
     | OnModuleOffsetLocationSequenceComponent
     | OnAddressableAreaOffsetLocationSequenceComponent
 )
 
+LabwareOffsetLocationSequenceComponents = Annotated[
+    LabwareOffsetLocationSequenceComponentsUnion, Field(discriminator="kind")
+]
+
```
Up to you, but if it makes things any cleaner, it would be safe to have a single type that is an `Annotated[Union[...], Field(discriminator=...)]` instead of splitting the `Annotated` from the `Union`.

sfoster1 replied:
I like having it separate because it makes it slightly easier for me to edit :)

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
```diff
@@ -0,0 +1,126 @@
+            sqlalchemy.insert(schema_10.labware_offset_table).values(
+                _v9_offset_to_v10_offset(offset)
+            )
+        ).lastrowid
```
Nit: The SQLAlchemy docs make [lastrowid](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.lastrowid) sound pretty low-level, and I find it confusing which ID it refers to. Would [inserted_primary_key](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.inserted_primary_key) work instead?

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
```diff
@@ -0,0 +1,126 @@
+        elif sequence_component.kind == "onAddressableArea":
+            primary_component_value = sequence_component.addressableAreaName
+            component_value_json = sequence_component.model_dump_json()
+        else:
+            # This should never happen since we're exhaustively checking kinds here
+            continue
```
Can this be `typing_extensions.assert_never(sequence_component.kind)` instead of `continue`?

sfoster1 replied:
as long as that doesn't result in a runtime assertion, yes. if it did, then this endpoint would 500 if we ever added a new component and then downmigrated, and i would love for that to not happen

sfoster1 replied:
which i'm realizing I did not explicitly handle and should, oops

SyntaxColoring commented on robot-server/robot_server/labware_offsets/router.py:
```diff
@@ -57,13 +56,13 @@ async def post_labware_offset(  # noqa: D103
-    request_body: Annotated[RequestModel[LegacyLabwareOffsetCreate], fastapi.Body()],
-) -> PydanticResponse[SimpleBody[LabwareOffset]]:
-    new_offset = LabwareOffset.model_construct(
+    request_body: Annotated[RequestModel[StoredLabwareOffsetCreate], fastapi.Body()],
```
Not for this PR, but if we're moving away from the "flat" way of representing locations, it seems like we (or I) need to rethink the query params in `GET /labwareOffsets` too, right? At this point, it seems like it should have a monolithic `locationSequence` query param that's just the JSON-encoded string of the model's `locationSequence` list.

sfoster1 replied:
yeah at some point we should rethink this. i think it probably is not that far off, though

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_10.py:
```diff
@@ -0,0 +1,420 @@
+    sqlalchemy.Column("created_at", UTCDateTime, nullable=False),
+)
+
+labware_offset_location_sequence_components_table = sqlalchemy.Table(
+    "labware_offset_sequence_components",
+    metadata,
+    # ID for this row, which largely won't be used
+    sqlalchemy.Column("row_id", sqlalchemy.Integer, primary_key=True),
+    # Which offset this belongs to
+    sqlalchemy.Column(
+        "offset_id",
+        sqlalchemy.ForeignKey(
+            "labware_offset_with_sequence.row_id",
+        ),
+        nullable=False,
+        index=True,
+    ),
+    # Its position within the sequence
+    sqlalchemy.Column("sequence_ordinal", sqlalchemy.Integer, nullable=False),
+    # An identifier for the component; in practice this will be an enum entry (of the kind values
+    # of the LabwareOffsetSequenceComponent models) but by keeping that out of the schema we don't
+    # have to change the schema if we add something new there
+    sqlalchemy.Column("component_kind", sqlalchemy.String, nullable=False),
+    # The value of the component, which will differ in kind by what component it is, and would be
+    # annoying to further schematize without yet more normalization. If we ever add a sequence component
+    # that has more than one value in it (think twice before doing this), pick a primary value that you'll
+    # be searching by and put that here.
+    sqlalchemy.Column("primary_component_value", sqlalchemy.String, nullable=False),
+    # If the value of the component has more than one thing in it, dump it to json and put it here.
+    sqlalchemy.Column("component_value_json", sqlalchemy.String, nullable=False),
+)
```
We can try this schema, but does it seem to you like it's panning out? The restriction on having a single `primary_component_value`...the complexity in `store.py` and `_search_query_builder.py` to implement, I surmise, [exact relational division](https://dba.stackexchange.com/q/161433/327987)...is this really better than just querying a monolithic JSON-encoded `location` column? Is this really faster for the machine to execute or easier for us to maintain?

I know JSON-encoded columns are a *generally* a red flag and the rule of thumb is *generally* to split things up. But also, the overriding rule is that the schema should support our actual query and update patterns. And it's not like we're doing anything that benefits from individual component access.

sfoster1 replied:
I think so, yeah. I'd rather have the searching done in sql; I think it will scale better to more offsets, and it's fundamentally the thing that sql is for. i'd bet there's a simpler way to do the query builder and am open to refactoring it in the future.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v9_to_v10.py:
```diff
@@ -0,0 +1,126 @@
+class Migration9to10(Migration):  # noqa: D101
+    def migrate(self, source_dir: Path, dest_dir: Path) -> None:
+        """Migrate the persistence directory from schema 9 to 10."""
+        # grab the deck def. middlewares aren't up yet so we can't use the nice version
```
Misplaced `# grab the deck def` comment?

SyntaxColoring commented on robot-server/robot_server/labware_offsets/store.py:
```diff
@@ -50,78 +36,71 @@ def __init__(self, sql_engine: sqlalchemy.engine.Engine) -> None:
+                sqlalchemy.insert(labware_offset_table).values(
+                    _pydantic_to_sql_offset(offset)
+                )
+            ).lastrowid
```
Same comment about `lastrowid`.

SyntaxColoring reviewed (approved):
Thanks!

I think we should take a hard look at whether we really need each location component to be its own row. It seems like it adds a lot of complexity, and I don't see what it gets us in return. In the same vein, also think we should consider overwriting the existing schema 9 instead of preserving it in the migration chain.

Besides that, various small things.

sfoster1 committed:
little fixups

sfoster1 committed:
review comments

sfoster1 committed:
fixup: do not crash when encountering an bad loc

This will allow us to add new kinds of location without doing schema
changes and migrations. If you add a location and downgrade to a version
that doesn't understand it, you'll get an
"UnknownLocationSequenceComponent", just like a bad run.
